File: services/api/src/api/schemas/align.py
from pydantic import BaseModel, ConfigDict


class AlignDomainDataPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    words: list[str]
    spaces: list[str]
    sent_ids: list[int]
    par_ids: list[int]
    sent_to_par_ids: dict[int, int]
    sent_to_word_ids: dict[int, list[int]]
    par_to_sent_ids: dict[int, list[int]]
    par_to_word_ids: dict[int, list[int]]

class AlignMapsPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    src_to_tgt: dict[int, list[int]]
    tgt_to_src: dict[int, list[int]]


class AlignRequestSentence(BaseModel):
    sent_id: int
    source: str
    target: str 

class AlignRequestParagraph(BaseModel):
    par_id: int
    sentences: list[AlignRequestSentence]

class AlignRequest(BaseModel):
    paragraphs: list[AlignRequestParagraph]
    src_lang: str
    tgt_lang: str

class AlignResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    src: AlignDomainDataPayload
    tgt: AlignDomainDataPayload
    align: AlignMapsPayload

File: apps/web/app/(protected)/documents/feature/types/align.ts
import type { AlignmentPayload } from "./pageTranslation"
import type { TranslateSentence, TranslateSentenceDTO, TranslateParagraph, TranslateParagraphDTO } from "./translate";

// -------------------------
// Frontend
// -------------------------
export type AlignRequest = {
  paragraphs: TranslateParagraph[];
  srcLang: string;
  tgtLang: string;
};

// -------------------------
// Backend return type
// -------------------------
export type AlignResponseDTO = {
  src: AlignmentPayload["src"];
  tgt: AlignmentPayload["tgt"];
  align: AlignmentPayload["align"];
};

export function toSentenceDTO(sent: TranslateSentence): TranslateSentenceDTO {
  return {
    sent_id: sent.sentId,
    source: sent.source,
    target: sent.target,
  };
}

export function toParagraphDTO(par: TranslateParagraph): TranslateParagraphDTO {
  return {
    par_id: par.parId,
    sentences: par.sentences.map(toSentenceDTO),
  };
}

export function toParagraphsDTO(
  paragraphs: TranslateParagraph[]
): TranslateParagraphDTO[] {
  return paragraphs.map(toParagraphDTO);
}